cmd/influxd/run.go

```diff
@@ -24,126 +26,219 @@ import (
 	"github.com/influxdb/influxdb/udp"
 )
 
-func Run(config *Config, join, version string) (*messaging.Broker, *influxdb.Server, *raft.Log) {
-	log.Printf("influxdb started, version %s, commit %s", version, commit)
+type RunCommand struct {
+	// The logger passed to the ticker during execution.
+	Logger    *log.Logger
+	logWriter *os.File
+	config    *Config
+	hostname  string
+	node      *Node
+}
+
+func NewRunCommand() *RunCommand {
+	return &RunCommand{
+		node: &Node{},
+	}
+}
 
-	var initBroker, initServer bool
-	if initBroker = !fileExists(config.BrokerDir()); initBroker {
-		log.Printf("Broker directory missing. Need to create a broker.")
+type Node struct {
+	broker   *influxdb.Broker
+	dataNode *influxdb.Server
+	raftLog  *raft.Log
+}
+
+func (s *Node) Close() {
+	if s.broker != nil {
+		if err := s.broker.Close(); err != nil {
+			log.Fatalf("error closing broker: %s", err)
+		}
 	}
 
-	if initServer = !fileExists(config.DataDir()); initServer {
-		log.Printf("Data directory missing. Need to create data directory.")
+	if s.raftLog != nil {
+		if err := s.raftLog.Close(); err != nil {
+			log.Fatalf("error closing raft log: %s", err)
+		}
 	}
-	initServer = initServer || initBroker
 
-	// Parse join urls from the --join flag.
-	var joinURLs []url.URL
-	if join == "" {
-		joinURLs = parseURLs(config.JoinURLs())
+	if s.dataNode != nil {
+		if err := s.dataNode.Close(); err != nil {
+			log.Fatalf("error data broker: %s", err)
+		}
+	}
+
+}
+
+func (cmd *RunCommand) Run(args ...string) error {
+	// Set up logger.
+	cmd.Logger = log.New(os.Stderr, "", log.LstdFlags)
+
+	// Parse command flags.
+	fs := flag.NewFlagSet("", flag.ExitOnError)
+	var configPath, pidfile, hostname, join, cpuprofile, memprofile string
+
+	fs.StringVar(&configPath, "config", "", "")
+	fs.StringVar(&pidfile, "pidfile", "", "")
+	fs.StringVar(&hostname, "hostname", "", "")
+	fs.StringVar(&join, "join", "", "")
+	fs.StringVar(&cpuprofile, "cpuprofile", "", "")
+	fs.StringVar(&memprofile, "memprofile", "", "")
+
+	fs.Usage = printRunUsage
+	fs.Parse(args)
+	cmd.hostname = hostname
+
+	// Start profiling, if set.
+	startProfiling(cpuprofile, memprofile)
+	defer stopProfiling()
+
+	// Print sweet InfluxDB logo and write the process id to file.
+	fmt.Print(logo)
+	writePIDFile(pidfile)
+
+	// Set parallelism.
+	runtime.GOMAXPROCS(runtime.NumCPU())
+	log.Printf("GOMAXPROCS set to %d", runtime.GOMAXPROCS(0))
+
+	var err error
+
+	// Parse configuration file from disk.
+	if configPath != "" {
+		cmd.config, err = parseConfig(configPath, hostname)
 	} else {
-		joinURLs = parseURLs(join)
+		cmd.config, err = NewTestConfig()
+	}
+
+	if err != nil {
+		cmd.Logger.Fatal(err)
+	} else if configPath == "" {
+		cmd.Logger.Println("No config provided, using default settings")
 	}
 
+	cmd.Open(cmd.config, join)
+
+	// Wait indefinitely.
+	<-(chan struct{})(nil)
+	return nil
+}
+
+func (cmd *RunCommand) Open(config *Config, join string) (*messaging.Broker, *influxdb.Server, *raft.Log) {
```

Comment:
I find this a little strange. I think of a "command" as something that tells the system to do something. But it's more than that here. This "command" has a `node`, it has an `Open` call, and which returns three different objects. Do I call `Open()` before I call `Run()`? The other way around?

I'm digging into this code, so perhaps the answer will be clearer, but the idea that a "command" has a `node` is definitely odd. Perhaps I am missing something, but I feel like it should be more like this:

`RunCommand(....the config options....) (*Node, err)`